integrations/google_ai/src/haystack_integrations/components/generators/google_ai/chat/gemini.py

```diff
@@ -297,98 +272,113 @@ def run(
             A list of `ChatMessage` instances, representing the input messages.
         :param streaming_callback:
             A callback function that is called when a new token is received from the stream.
+        :param tools:
+            A list of tools for which the model can prepare calls. If set, it will override the `tools` parameter set
+            during component initialization.
         :returns:
             A dictionary containing the following key:
             - `replies`:  A list containing the generated responses as `ChatMessage` instances.
         """
         streaming_callback = streaming_callback or self._streaming_callback
-        history = [self._message_to_content(m) for m in messages[:-1]]
-        session = self._model.start_chat(history=history)
 
-        new_message = self._message_to_part(messages[-1])
+        tools = tools or self._tools
+        _check_duplicate_tool_names(tools)
+        google_tools = [self._convert_tool_to_google_tool(tool) for tool in tools] if tools else None
+
+        if messages[0].is_from("system"):
+            self._model._system_instruction = content_types.to_content(messages[0].text)
+            messages = messages[1:]
+
+        google_messages = [_convert_chatmessage_to_google_content(m) for m in messages]
+
+        session = self._model.start_chat(history=google_messages[:-1])
```

Comment:
Whaaat, they have sessions? The must be keeping these on client in some internal structures?

Reply:
TBH, I don't know.
This API is really hard to navigate...